src/python/aqueduct_executor/operators/connectors/tabular/s3.py

```diff
@@ -29,58 +36,121 @@ def authenticate(self) -> None:
     def discover(self) -> List[str]:
         raise Exception("Discover is not supported for S3.")
 
-    def _fetch_object(self, key: str, format: common.S3FileFormat) -> pd.DataFrame:
+    def _fetch_object(self, key: str, params: extract.S3Params) -> Any:
         response = self.s3.Object(self.bucket, key).get()
         data = response["Body"].read()
-        buf = io.BytesIO(data)
-        if format == common.S3FileFormat.CSV:
-            return pd.read_csv(buf)
-        elif format == common.S3FileFormat.JSON:
-            return pd.read_json(buf)
-        elif format == common.S3FileFormat.PARQUET:
-            return pd.read_parquet(buf)
+        if params.artifact_type == ArtifactType.TABULAR:
+            if params.format is None:
+                raise Exception("You must specify a file format for tabular data.")
+            buf = io.BytesIO(data)
+            if params.format == common.S3TabularFormat.CSV:
+                return pd.read_csv(buf)
+            elif params.format == common.S3TabularFormat.JSON:
+                return pd.read_json(buf)
+            elif params.format == common.S3TabularFormat.PARQUET:
+                return pd.read_parquet(buf)
+            raise Exception("Unknown S3 file format %s." % params.format)
+        elif params.artifact_type == ArtifactType.JSON:
+            # This assumes that the encoding is "utf-8". May worth considering letting the user
+            # specify custom encoding in the future.
+            json_data = data.decode(_DEFAULT_JSON_ENCODING)
+            # Make sure the data is a valid json object.
+            json.loads(json_data)
```

Comment:
we should probably raise a nicer error message if the data is not json - make it really clear to them that the data in their S3 is not in json format.

Reply:
Let me check what the error message is if we don't do anything. If it's confusing I can add a catch to raise a nicer one.